File: fractizer.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdint.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>

typedef struct fparams_t {
    uint16_t max_iters;
    double   escape_val;
    double   x_min;
    double   x_max;
    double   y_min;
    double   y_max;
    uint16_t x_pels;
    uint16_t y_pels;
    uint16_t x_tile;
    uint16_t y_tile;
    uint8_t  type;
} fparams_t;


fparams_t get_params(int argc, char *argv[]) {
    fparams_t params;
    params.max_iters = 200;
    params.escape_val = 4;
    params.x_min = -2;
    params.x_max = 2;
    params.y_min = -2;
    params.y_max = 2;
    params.x_pels = 1024;
    params.y_pels = 1024;
    params.x_tile = 0;
    params.y_tile = 0;
    params.type = 0;

    if (argc > 1)  params.max_iters   = atol(argv[1]);
    if (argc > 2)  params.escape_val  = atof(argv[2]);
    if (argc > 3)  params.x_min       = atof(argv[3]);
    if (argc > 4)  params.x_max       = atof(argv[4]);
    if (argc > 5)  params.y_min       = atof(argv[5]);
    if (argc > 6)  params.y_max       = atof(argv[6]);
    if (argc > 7)  params.x_pels      = atol(argv[7]);
    if (argc > 8)  params.y_pels      = atol(argv[8]);
    if (argc > 9)  params.x_tile      = atol(argv[9]);
    if (argc > 10) params.y_tile      = atol(argv[10]);
    if (argc > 11) params.type        = atol(argv[11]);
    return params;

};

/*
int open_socket(char *host, uint16_t port) {
    int sockfd = 0;
    int n =0;
    struct sockaddr_in serv_addr;

    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        puts("Could not create socket.");
        return sockfd;
    }
    memset(&serv_addr, '0', sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(port);

    int pt_res = inet_pton(AF_INET, host, &serv_addr.sin_addr);
    if (pt_res <= 0) {
        printf("HOST: %s pt_res %d\n",host, pt_res);
        puts("inet_pton error");
        return -1;
    }

    if (connect(sockfd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0) {
        puts("connect failed");
        return -1;
    }
    return sockfd;
};
*/


static inline uint8_t calc_pixel_mb(fparams_t *pparams, double x, double y, uint16_t *piters, double *pval) {
    double z_r = 0;
    double z_i = 0;
    double c_r = x;
    double c_i = y;
    uint16_t iter = 0;
    double val = 0;
    while ((iter < pparams->max_iters) && (val < pparams->escape_val)) {
        double z_n1_r = z_r * z_r - z_i * z_i + c_r;
        double z_n1_i = 2.0 * z_r * z_i + c_i;
        val = sqrt(z_n1_r * z_n1_r + z_n1_i * z_n1_i);
        iter++;
        z_r = z_n1_r;
        z_i = z_n1_i;
    }
    *pval = val;
    *piters = iter;
    return iter < pparams->max_iters;
};

static inline uint8_t calc_pixel_mbbs(fparams_t *pparams, double x, double y, uint16_t *piters, double *pval) {
    double z_r = 0;
    double z_i = 0;
    double c_r = x;
    double c_i = y;
    uint16_t iter = 0;
    double val = 0;
    while ((iter < pparams->max_iters) && (val < pparams->escape_val)) {
        z_r = fabs(z_r);
        z_i = fabs(z_i);
        double z_n1_r = z_r * z_r - z_i * z_i + c_r;
        double z_n1_i = 2.0 * z_r * z_i + c_i;
        val = sqrt(z_n1_r * z_n1_r + z_n1_i * z_n1_i);
        iter++;
        z_r = z_n1_r;
        z_i = z_n1_i;
    }
    *pval = val;
    *piters = iter;
    return iter < pparams->max_iters;
};

static inline uint8_t calc_pixel_mb3(fparams_t *pparams, double x, double y, uint16_t *piters, double *pval) {
    double z_r = 0;
    double z_i = 0;
    double c_r = x;
    double c_i = y;
    uint16_t iter = 0;
    double val = 0;
    while ((iter < pparams->max_iters) && (val < pparams->escape_val)) {
        double z_n1_r = z_r * z_r * z_r - 3 * z_r * z_i * z_i + c_r;
        double z_n1_i = 3 * z_r * z_r * z_i - z_i * z_i * z_i + c_i;
        val = sqrt(z_n1_r * z_n1_r + z_n1_i * z_n1_i);
        iter++;
        z_r = z_n1_r;
        z_i = z_n1_i;
    }
    *pval = val;
    *piters = iter;
    return iter < pparams->max_iters;
};

void generate_fractal(fparams_t *pparams, uint16_t *rbuf) {
    double x_step = (pparams->x_max - pparams->x_min) / (double)pparams->x_pels;
    double y_step = (pparams->y_max - pparams->y_min) / (double)pparams->y_pels;
    for (uint16_t j=0; j<pparams->y_pels; j++) {
        double y = pparams->y_min + j * y_step;
        for (uint16_t i=0; i<pparams->x_pels; i++) {
            double x = pparams->x_min + i * x_step;
            uint16_t iters = 0;
            double val = 0;
            switch (pparams->type) {
                case 1 : calc_pixel_mb3(pparams, x,y, &iters, &val); break;
                case 2 : calc_pixel_mbbs(pparams, x,y, &iters, &val); break;
                default: calc_pixel_mb(pparams, x,y, &iters, &val);
            }
            rbuf[i + j * pparams->x_pels] = iters;
        }
    };
};

void dump_fractal(fparams_t *pparams, uint16_t *rbuf) {
    printf("%u %f %f %f %f %f %u %u %u %u %u\n",
            pparams->max_iters,
            pparams->escape_val,
            pparams->x_min,
            pparams->x_max,
            pparams->y_min,
            pparams->y_max,
            pparams->x_pels,
            pparams->y_pels,
            pparams->x_tile,
            pparams->y_tile,
            pparams->type
            );
    uint32_t len = pparams->x_pels * pparams->y_pels;
    for (uint32_t i=0; i<len; i++) {
        printf("%u\n",rbuf[i]);
    }
    printf("all_done\n");
    fflush(stdout);
};

int main(int argc, char *argv[]) {
    fparams_t params = get_params(argc, argv);

    uint16_t *rdata = malloc(sizeof(uint16_t) * params.x_pels * params.y_pels);
    if (rdata) {
        generate_fractal(&params, rdata);
        if (1) {
            dump_fractal(&params, rdata);
        }
    }
};

File: fractizer.js

function cmag(z)    { return Math.sqrt(z.r*z.r + z.i*z.i); }
function csmul(z,s) { return { r: z.r * s, i: z.i * s };   }
function ccube(z)   {
    return {
        r: z.r*z.r*z.r - 3*z.r*z.i*z.i,
        i: 3*z.r*z.r*z.i - z.i*z.i*z.i,
    };
}
function cadd(z,y)  {
    return {
        r: z.r + y.r,
        i: z.i + y.i,
    };
}
function csquare(z) {
    return {
        r: z.r*z.r - z.i*z.i,
        i: 2*z.r*z.i,
    };
}

function znp1_mb(z,c) {
    var nzr = csquare(z);
    return cadd(nzr,c);
}
function znp1_cos(z,c) {
    var nz = { r:0, i:0 };
    nz.r = Math.cos(z.r) * Math.cosh(z.i) + c.r;
    nz.i = Math.sin(z.r) * Math.sinh(z.i) + c.i;
    return nz;
}
function znp1_mbtc(z,c) {
    z.i = -z.i;
    return cadd(csquare(z),c);
}
function znp1_mbbs(z,c) {
    z.r = Math.abs(z.r);
    z.i = Math.abs(z.i);
    return cadd(csquare(z),c);
}
function znp1_mb3(z,c) {
    return cadd(ccube(z),c);
}
function znp1_nf3(z,c) {
    var nz = ccube(z);
    nz.r += c.r - 1.0;
    nz.i += c.i;
    return nz;
}
function znp1_nf3m2z(z,c) {
    return cadd(cadd(ccube(z),csmul(z,-2)),c);
}
function znp1_nf8p15z4(z,c) {
    var z2 = csquare(z);
    var z4 = csquare(z2);
    var z8 = csquare(z4);
    var z14_15 = csmul(z4,15);
    var nz = cadd(z8,z14_15);
    return cadd(nz,c);
}

function calc_pixel_generic(args,x,y) {
    c = { r:0, i: 0};
    z = { r:0, i: 0};
    if (args.do_julia) {
        z.r = x; z.i = y; 
        c.r = args.jx; c.i = params.jy;
    } else {
        z.r = 0; z.i = 0;
        c.r = x; c.i = y;
    }
    var iter = 0;
    var val = 0;
    while ((iter < args.max_iters) && (val < params.escape_val)) {
        var nz = args.algo(z,c);
        val = cmag(nz);
        z = nz;
        iter++;
    }
    return iter;
}

function generate_fractal(args) {
    console.log('generate_fractal()');
    console.log(args);
    var x_step = (args.x_max - args.x_min) / args.x_pels;
    var y_step = (args.y_max - args.y_min) / args.y_pels;

    var data = new Array(args.x_pels * args.y_pels);
    var algo = znp1_mb;

    switch (args.type) {
        case 1 : algo = znp1_mb3; break;
        case 2 : algo = znp1_mbbs; break;
        case 3 : algo = znp1_mbtc; break;
        case 4 : algo = znp1_cos; break;
        case 5 : algo = znp1_nf3; break;
        case 6 : algo = znp1_nf3m2z; break;
        case 7 : algo = znp1_nf8p15z4; break;
        default: algo = znp1_mb; break;
    }

    args.algo = algo;

    for (var j=0; j<args.y_pels; j++) {
        var y = args.y_min + j * y_step;
        for (var i=0; i<args.x_pels; i++) {
            var x = args.x_min + i * x_step;
            var iters = 0;
            var val = 0;
            iters = calc_pixel_generic(args, x, y);
            data[i + j * args.x_pels] = iters;
        }
    }
    console.log('generate_done');
    return data;
}

